### ISSUE TITLE
Throwing an exception instead of an empty result makes rich features difficult to maintain

### ISSUE THREAD

**Issue by plarger-eptica** [pac4j/pac4j#1072]
Hi,

In order to support heavy and dynamic multi-tenancy and because of the way the _Clients.findClient_ or _ClientFinder.find_ methods are implemented, I found myself writing things like this:
```
try{
    return findClientInOneSource(context);
} catch (TechnicalException e){
    return findClientInOtherSource(context);
}
```
This troubles me because:
- I had to read the source code to understand this
- if at any moment another TechnicalException is thrown for another reason, my code will misbehave,
- it makes it difficult to handle multiple sources of clients (try/catch in a loop?),
- throwing an exception is an expensive operation for such a case,

IMHO, the decision of throwing an exception should be left to the class exploiting the _Clients.findClient_ or _ClientFinder.find_ methods. Returning an Optional or an empty list makes much more sense to me.

I take this issue as an opportunity to thank you for your work, Pac4J has been of a great help to us!

**Comment by leleuj** [comment]
That's an interesting feedback! My feeling is that you should create your own `CustomClients` and use it instead of the default `Clients` to handle your specific use cases.

That said, we're working on using `Optional`s in the library, so we could change that. Do you mean we should keep the actual behaviour (throw an exception when the client is not found) or do you have something else in mind?

**Comment by pierrelarger** [comment]
Hi, still plarger-eptica here, sorry about the delay.
I kind of like seeing a stack trace in my log file in this case, it clearly shows a misuse of Pac4J. What bothers me is that a class/method whose role is to find clients also has the role of saying that finding at least once is mandatory. I think the latter should be done in the _DefaultSecurityLogic_ or something like that. This feels less rigid and more SOLID-compliant to me. In fact, I did encounter this when implementing my own _Customclients_ and it felt like an avoidable source of complexity.

I have to admit my own code will break if my suggestion is applied but I am convinced that changing this would make it easier for a newcomer to implement something like I did (dynamic multi-tenancy).

I can submit a pull request or a patch in a few days if you want.
